### PR TITLE
Type safe API, significantly simplify Core, Remove Capture, Update to 3.15, Split out features into modules, Use TH to generate datatypes, Add proper support for dynamic registration

### PR DESCRIPTION
**Pull request by wz1000**
This PR is a big, breaking change that has spent a long time in the oven. I am requesting reviews while I work to patch upstream packages.

The main motivation behind this is to provide a safe API for constructing and sending LSP requests and responses. The `method` field of request messages is no longer divorced from the `param` field,
so as to statically ensure that you provide the correct payload for your messages

The methods are represented on the type level by a GADT tagged by the methods provenance(`FromServer | FromClient`), and its message type (`Request | Notification`). At the value level, it is represented by a constructor of a singleton type for the method.

```haskell
data Method (p :: Provenance) (t :: MethodType) where
  Initialize  :: Method FromClient Request
  Initialized :: Method FromClient Notification
  ...
data SMethod (m :: Method p t) where
  SInitialize  :: SMethod Initialize
  SInitialized :: SMethod Initialized
  ...
```

These methods are mapped to parameter and response types with the help of two type families,

```haskell
type family MessageParams (m :: Method p t) :: Type where
  MessageParams Initialize  = InitializeParams
  MessageParams Initialized  = Maybe InitializedParams
  ...
type family ResponseParams (m :: Method p Request) :: Type where
  ResponseParams Initialize = InitializeResponseCapabilities
  ResponseParams Shutdown   = Empty
  ...
```

Using this encoding, you only need to check these two places to know everything
you need about a particular method type

As these are all listed in order and grouped by provenance, correlating with the LSP spec becomes easy.

We ensure the well-formedness of responses by tagging `LspId` with a phantom
method type parameter, which is used in `ResponseMessage` to ensure that
you are responding to a request with the correct params.

Major changes in `haskell-lsp-types`:
- Method definitions moved to `Language.Haskell.LSP.Types.Method`
- Definition of `RequestMessage`, `NotificationMessage` and `ResponseMessage` updated using the `MessageParams` and `ResponseParams` type families in order to be typesafe as described above
- `LspId` moved to `Language.Haskell.LSP.Types.LspId`
- Type synonyms moved to new module `Language.Haskell.LSP.Types.Synonyms`
- New module `Data.IxMap` that helps in dealing with method-tagged ids and response handlers in a type-safe manner 

Major changes in `haskell-lsp`:
- Deleted `Capture.hs`
  @bubba and I came to this conclusion after discussing the state of `lsp-test`. 
  The capturing mechanism in `haskell-lsp` is currently lacking in quite a few ways and even misses some communication with the client.
  The plan is to add a `lsp-test-conduit` tool to lsp-test that wraps around language servers and accurately captures
  information independent of `haskell-lsp`
- Deleted `Messages.hs`: it is no longer required
- Significantly reworked Core.hs
  * The handlerMap mechanism was simplified:
      + The `Handlers` type was replaced by a function `J.SClientMethod m -> Maybe (Handler m)`
      + `hh` was slightly altered in order to be generalizable for all requests
      + `handlerMap` now calls `hh` to handle all request. Any special cases/extra work to be done by `haskell-lsp` for this request
        is uniformly handled by generalizing the `changeVFS` parameter that `hh` previously accepted.
  * The `sendFunc` in `LspFuncs` was changed to be the typesafe `ServerMessageFunc`
    `ServerMessageFunc` is `SomeServerMessageWithResponse -> IO ()`
    `SomeServerMessageWithResponse` is a type safe way to specify both a message to
    be sent from the server to the client as well as forcing you to specify
    a handler/callback for the eventual response
  * In order send responses to client messages, we include a
  `ClientResponseHanlder` callback with every message that you can call with a
  (well-typed) reponse to the message

Suggested reading order:  

1. `Language/Haskell/LSP/Types/Method.hs `
2. `Language/Haskell/LSP/Types/Message.hs`
3. `Language/Haskell/LSP/Types/LspId.hs`

These should give you a basic overview of what this patch is trying to accomplish

After this, the changes in `Language/Haskell/LSP/Core.hs` are pretty significant

Lastly, the following modules contain some new utility code:

- `Data/IxMap.hs`
- `Language/Haskell/LSP/Types/Utils.hs`

The rest of the changes are pretty trivial.

TODO:

- Update tests
- Rework `lsp-test`
- Patch upstream libraries

### CODE REVIEW

**Comment by bubba** [comment]
Great stuff! Going to try plugging this into lsp-test to dogfood this. At the very least this will require a major major version bump (i.e. haskell-lsp-types-1.0) or putting this into alternative library a-la haskell-lsp-types-safe if this type level stuff is too much overhead for users.

**Comment by bubba** [comment]
I'm currently working on migrating the example server

**Comment by bubba** [comment]
- [ ] Remove handler for initialization and provide a different callback, since haskell-lsp modifies it anyway

File: src/Language/Haskell/LSP/Core.hs
Comment (wz1000) [inline src/Language/Haskell/LSP/Core.hs:706]: Couldn't these notifications be moved inside the `Right () -> ` case branch above to fix the issue?

Comment (bubba) [reply src/Language/Haskell/LSP/Core.hs:706]: Unfortunately it needs to return the result of running the user's callback, which would have to happen inside the response handler. I think it might be possible with the use of some async

**Comment by bubba** [comment]
- [ ] Make [`Registration`](https://hackage.haskell.org/package/haskell-lsp-types-0.22.0.0/docs/Language-Haskell-LSP-Types.html#t:Registration)'s registerOptions field dependent on the ClientMethod

**Comment by alanz** [comment]
One high-level comment: this PR has

```haskell
data Method (p :: Provenance) (t :: MethodType) where
```

It is not immediately clear what `Provenance` is. Perhaps consider renaming it to something obvious, like `From`?

File: src/Language/Haskell/LSP/Core.hs
Comment (Sir4ur0n) [inline src/Language/Haskell/LSP/Core.hs:474]: Is this a question for maintainers? Or just a reminder to solve this before merging?

**Review by Sir4ur0n — COMMENTED** [review]
Granted, this may look weird for a total stranger to review this request 😅 

Actually I was about to start working on a PR to add some documentation on this repo (because when I first tried to hack on HLS a few weeks ago, I got quickly frustrated by the lack of documentation on LSP types) and before starting, I checked open PRs (to ensure this was not already worked on) and I noticed this HUGE PR, which may very well induce many conflicts with any PR I could start hacking.

So long story short, I decided to review this PR, with 0 knowledge of this repo 😁 

As such, my questions/remarks are mostly about documentation and style, I hope you don't mind 🙇 

Cheers and good luck with this huge PR!

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs
Comment (wz1000) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs:450]: It can only be a custom method.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs
Comment (wz1000) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Message.hs]: Yes, thanks

File: haskell-lsp-types/src/Data/IxMap.hs
Comment (bubba) [reply haskell-lsp-types/src/Data/IxMap.hs:7]: This module should be internal to haskell-lsp and haskell-lsp-types, maybe we should move it to `Language.Haskell.LSP.Types.Internal.IxMap` or something

Comment (wz1000) [reply haskell-lsp-types/src/Data/IxMap.hs:7]: I think it might be useful if users want to make maps keyed on `LspId` or `RegistrationId`.

File: haskell-lsp-types/src/Language/Haskell/LSP/Types/Location.hs
Comment (bubba) [reply haskell-lsp-types/src/Language/Haskell/LSP/Types/Location.hs:13]: The inline comments were out of date for the most part unfortunately. What I would like to ultimately see however is the specification documentation as haddocks on the types themselves (which we might need https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3330 first!)

**Review by bubba — APPROVED** [review]

**Pull request merged**
